File: packages/dotnet/src/Yourdudeken.Mpesa/Services/AccountService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Yourdudeken.Mpesa.Config;
using Yourdudeken.Mpesa.Http;
using Yourdudeken.Mpesa.Auth;

namespace Yourdudeken.Mpesa.Services
{
    public class AccountService
    {
        private readonly HttpClientWrapper _httpClient;
        private readonly AuthService _auth;
        private readonly MpesaConfig _config;
        private readonly string _baseUrl;

        public AccountService(HttpClientWrapper httpClient, AuthService auth, MpesaConfig config)
        {
            _httpClient = httpClient;
            _auth = auth;
            _config = config;
            _baseUrl = httpClient.GetBaseUrl();
        }

        public async Task<string> Balance(string shortcode, int identifierType, string remarks,
            string resultUrl = null, string timeoutUrl = null, string shortCodeType = "C2B")
        {
            var url = $"{_baseUrl}/mpesa/accountbalance/v1/query";
            var body = new Dictionary<string, object>
            {
                { "Initiator", _config.InitiatorName },
                { "SecurityCredential", GenerateSecurityCredential() },
                { "CommandID", "AccountBalance" },
                { "PartyA", shortcode },
                { "IdentifierType", identifierType },
                { "Remarks", remarks },
                { "ResultURL", resultUrl ?? _config.BalanceResultUrl },
                { "QueueTimeOutURL", timeoutUrl ?? _config.BalanceTimeoutUrl }
            };

            var token = await _auth.GetAccessToken(shortCodeType);
            return await _httpClient.Post(url, body, token);
        }

        public async Task<string> Status(string shortcode, string transactionId, int identifierType, string remarks,
            string resultUrl = null, string timeoutUrl = null, string shortCodeType = "C2B")
        {
            var url = $"{_baseUrl}/mpesa/transactionstatus/v1/query";
            var body = new Dictionary<string, object>
            {
                { "Initiator", _config.InitiatorName },
                { "SecurityCredential", GenerateSecurityCredential() },
                { "CommandID", "TransactionStatusQuery" },
                { "TransactionID", transactionId },
                { "PartyA", shortcode },
                { "IdentifierType", identifierType },
                { "Remarks", remarks },
                { "Occassion", "" },
                { "ResultURL", resultUrl ?? _config.StatusResultUrl },
                { "QueueTimeOutURL", timeoutUrl ?? _config.StatusTimeoutUrl }
            };

            var token = await _auth.GetAccessToken(shortCodeType);
            return await _httpClient.Post(url, body, token);
        }

        public async Task<string> Reversal(string shortcode, string transactionId, double amount, string remarks,
            string resultUrl = null, string timeoutUrl = null, string shortCodeType = "C2B")
        {
            var url = $"{_baseUrl}/mpesa/reversal/v1/request";
            var body = new Dictionary<string, object>
            {
                { "Initiator", _config.InitiatorName },
                { "SecurityCredential", GenerateSecurityCredential() },
                { "CommandID", "TransactionReversal" },
                { "TransactionID", transactionId },
                { "Amount", amount },
                { "ReceiverParty", shortcode },
                { "RecieverIdentifierType", "11" },
                { "Remarks", remarks },
                { "Occasion", "" },
                { "ResultURL", resultUrl ?? _config.ReversalResultUrl },
                { "QueueTimeOutURL", timeoutUrl ?? _config.ReversalTimeoutUrl }
            };

            var token = await _auth.GetAccessToken(shortCodeType);
            return await _httpClient.Post(url, body, token);
        }

        private string GenerateSecurityCredential() => "";
    }
}

File: packages/java/src/main/java/com/yourdudeken/mpesa/services/AccountService.java
package com.yourdudeken.mpesa.services;

import com.yourdudeken.mpesa.config.MpesaConfig;
import com.yourdudeken.mpesa.http.HttpClient;
import com.yourdudeken.mpesa.auth.AuthService;
import java.util.HashMap;
import java.util.Map;

public class AccountService {
    private HttpClient httpClient;
    private AuthService auth;
    private MpesaConfig config;
    private String baseUrl;

    public AccountService(HttpClient httpClient, AuthService auth, MpesaConfig config) {
        this.httpClient = httpClient;
        this.auth = auth;
        this.config = config;
        this.baseUrl = httpClient.getBaseUrl();
    }

    public String balance(String shortcode, int identifierType, String remarks,
                         String resultUrl, String timeoutUrl, String shortCodeType) {
        String url = baseUrl + "/mpesa/accountbalance/v1/query";
        Map<String, Object> body = new HashMap<>();
        body.put("Initiator", config.getInitiatorName());
        body.put("SecurityCredential", generateSecurityCredential());
        body.put("CommandID", "AccountBalance");
        body.put("PartyA", shortcode);
        body.put("IdentifierType", identifierType);
        body.put("Remarks", remarks);
        body.put("ResultURL", resultUrl != null ? resultUrl : config.getBalanceResultUrl());
        body.put("QueueTimeOutURL", timeoutUrl != null ? timeoutUrl : config.getBalanceTimeoutUrl());

        String token = auth.getAccessToken(shortCodeType);
        return httpClient.post(url, body, token);
    }

    public String status(String shortcode, String transactionId, int identifierType, String remarks,
                        String resultUrl, String timeoutUrl, String shortCodeType) {
        String url = baseUrl + "/mpesa/transactionstatus/v1/query";
        Map<String, Object> body = new HashMap<>();
        body.put("Initiator", config.getInitiatorName());
        body.put("SecurityCredential", generateSecurityCredential());
        body.put("CommandID", "TransactionStatusQuery");
        body.put("TransactionID", transactionId);
        body.put("PartyA", shortcode);
        body.put("IdentifierType", identifierType);
        body.put("Remarks", remarks);
        body.put("Occassion", "");
        body.put("ResultURL", resultUrl != null ? resultUrl : config.getStatusResultUrl());
        body.put("QueueTimeOutURL", timeoutUrl != null ? timeoutUrl : config.getStatusTimeoutUrl());

        String token = auth.getAccessToken(shortCodeType);
        return httpClient.post(url, body, token);
    }

    public String reversal(String shortcode, String transactionId, double amount, String remarks,
                           String resultUrl, String timeoutUrl, String shortCodeType) {
        String url = baseUrl + "/mpesa/reversal/v1/request";
        Map<String, Object> body = new HashMap<>();
        body.put("Initiator", config.getInitiatorName());
        body.put("SecurityCredential", generateSecurityCredential());
        body.put("CommandID", "TransactionReversal");
        body.put("TransactionID", transactionId);
        body.put("Amount", amount);
        body.put("ReceiverParty", shortcode);
        body.put("RecieverIdentifierType", "11");
        body.put("Remarks", remarks);
        body.put("Occasion", "");
        body.put("ResultURL", resultUrl != null ? resultUrl : config.getReversalResultUrl());
        body.put("QueueTimeOutURL", timeoutUrl != null ? timeoutUrl : config.getReversalTimeoutUrl());

        String token = auth.getAccessToken(shortCodeType);
        return httpClient.post(url, body, token);
    }

    private String generateSecurityCredential() {
        return "";
    }
}
